BEPs/BEP-396.md

```diff
@@ -0,0 +1,187 @@
+
+However, these additional retries waste a lot of computing resources. The read-write set generated by transaction execution can easily build a dependency graph between transactions, called `TxDAG`. This data can be used in the validating block phase to accelerate block execution and further improve the overall performance of the blockchain.
+
+## 5. Specification
+
+In BSC, the block interval is 3s, that is, the more transactions are executed within 3s, the better the chain performance and the greater the throughput.
+
+The two parts that take the most time in the 3s are the validating block and the mining block. Validators often validate the correctness of the current block and continue to mine the next block based on the latest state.
+
+To improve the performance of block execution, transactions need to be executed in parallel as much as possible to make full use of computing and IO resources. Imagine that if the dependencies of transactions are known, then the parallel execution path can be directly determined, but the dependencies need to be obtained by the read-write set after the transaction pre-execution (static analysis is not within the scope of this BEP).
+
+So two different TX execution strategies can be identified:
+
+![](./assets/BEP-396/pic1.png)
+
+In the mining phase, TXs of blocks are executed in parallel or serial to generate `TxDAG`. In the validating phase, `TxDAG` is used to further improve the parallel execution efficiency.
+
+### 5.1 Tx Dependency
+
+Assume that the execution of each transaction is atomic, and the input and output of the execution generate a state read-write set. In concurrent processing, read-only access to the resources is safe, but write operations require exclusive resources.
+
+At the same time, all transactions in the blockchain have a preset order, and transactions with lower TxIndex have higher execution priority.
+
+Therefore, the dependency constraint goes to, if transaction Txi writes `StateA` first, then transaction Txj reads `StateA` and j>i is satisfied. That is to say, Txj depends on Txi, Txj must be executed after Txi.
+
+If a transaction reads and writes the same value at a memory location, the location can be considered read-only.
+
+### 5.2 TxDAG Structure
+
+The structure of `TxDAG` is defined as follows.
+
+```go
+type TxDep struct {
+	// It describes the Relation with below txs 
+	// 0: this tx depends on below txs 
+	// 1: this transaction does not depend on below txs, all other previous txs depend on
+	Relation  uint8
+	TxIndexes []int
+}
+
+type TxDAG struct {
+	// The TxDAG type
+	// 0: delay the distribution of GasFee, it will ignore all gas fee distribution when tx execute
+	// 1: timely distribution of transaction fees, it will keep partial serial execution when tx cannot delay the distribution
+	Type uint8
+	// Tx Dependency List, the list index is equal to TxIndex
+	TxDeps []TxDep
+	// It indicates the scheduling priority of the transactions
+	SchedulePriority []int
```

Comment:
I dont think a validator should assign the `SchedulePriority` to all other nodes in the network. The other fullnode can decide the `Priority` by themselves according to the `TxDeps`.  For example, it will enforce other other client like Reth to implement exact same `Scheduler` that Geth implemented, I think the decoupling is better.

Reply:
Yeah, the priority queue of TxDAG is better to be determined, but the description here is vague. I originally planned to use the accumulated Gas of the execution path as the weight, but it is still under discussion. Another more compromise method is to remove the queue first, which is decided by the consumer as you said.